source/extensions/filters/udp/dns_filter/dns_filter_resolver.cc

```diff
@@ -0,0 +1,116 @@
+#include "extensions/filters/udp/dns_filter/dns_filter_resolver.h"
+
+#include "common/network/utility.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace UdpFilters {
+namespace DnsFilter {
+
+void DnsFilterResolver::resolveExternalQuery(DnsQueryContextPtr context,
+                                             const DnsQueryRecord* domain_query) {
+  // Create an external resolution context for the query.
+  LookupContext ctx{};
+  ctx.query_rec = domain_query;
+  ctx.query_context = std::move(context);
+  ctx.expiry = std::chrono::duration_cast<std::chrono::seconds>(
+                   dispatcher_.timeSource().systemTime().time_since_epoch())
+                   .count() +
+               std::chrono::duration_cast<std::chrono::seconds>(timeout_).count();
+  ctx.resolver_status = DnsFilterResolverStatus::Pending;
+
+  timeout_timer_->disableTimer();
```

Comment:
This is problematic for multiple reasons:
1) You early return below in the default case, potentially disabling the timer for other pending queries.
2) If you get a steady trickle of requests, you will forever be extending the timer.

If you are going to use a single timer, you will need some more intelligent code to make sure you handle both of the above cases. My advice for now is to just use a timer per-query and optimize it later as the code will be a lot simpler.

Reply:
OK.  This confirms a suspicion I had.  I will rework this.   My concern was that I'd be allocating a timer per query.  In any case, I'll update.

Reply:
This is still an issue, right? Please ping back when ready for review again.

/wait

Reply:
Ugh, apparently I overwrote my changes at some point.  I spent a bit of time fixing the mocking and expectations in the test.  I will update this....again.  I wish I could use your avatar as an emoji right now.

Reply:
@mattklein123 This change has been done.  Please review again when you have the cycles.   The CI failure does not appear to be filter related and a similar failure appears on other PRs.  I can monitor master for a fix and merge again.